docs/available-filters/envelope-encryption.adoc

```diff
@@ -0,0 +1,185 @@
+:kms-api-javadoc: https://javadoc.io/doc/io.kroxylicious/kroxylicious-kms/latest
+:encryption-api-javadoc: https://javadoc.io/doc/io.kroxylicious/kroxylicious-encryption
+
+= Envelope Encryption
+
+== What is it?
+
+A filter that provides a transparent https://kroxylicious.io/use-cases/[encryption-at-rest solution] for Apache Kafka.
+
+It works as follows:
+
+1. The filter intercepts the requests sent from the producing Kafka Client containing records.  The records
+   contained with the request are encrypted.  This happens before the records reach the Kafka Broker.
+2. The filter intercepts the responses sent from the Kafka Broker to the consuming Kafka Client.  The records
+   contained within the response are decrypted.  This happens before the records reach Kafka Client.
+
+The entire process is transparent from the point of view of both Kafka Client and Kafka Broker.  Neither are
+aware that the records are being encrypted.  Neither client nor broker have any access to the encryption keys.
+
+The filter encrypts the kafka records using a symmetric encryption key. It is possible to use different encryption
+keys for different topics. The encryption keys are held within by a Key Management Service (KMS).  It is the user's
+responsibility to provide the KMS service and pre-provision the encryption keys within it.
+
+== Prerequisites
+
+In order to use the filter, a compatible KMS (Key Management Service) must be provided.  The KMS is used to store
+the encryption keys used to encrypt the records sent to the topic.  The service is used at runtime. If the KMS is
+unavailable, it will become impossible to produce or consume records through the filter until the KMS service is
+restored.  It is recommended to use a KMS in a highly available configuration.
+
+Currently, there is a single KMS integration: https://www.hashicorp.com/[HashiCorp Vault&#174;].  More KMS integrations
+are planned soon.
+
+It is also possible for a user to implement their own KMS integration.  The implementation must implement the
+{kms-api-javadoc}/io/kroxylicious/kms/service/KmsService.html[KMS public API] and make the implementation available on
+the classpath with the service loader mechanism.
+
+== How to use the filter
+
+There are three steps to using the filter.
+
+1. Setting up the KMS.
+2. Configuring the filter within Kroxylicious.
```

Comment:
Do we need to talk about setting up the classpath? The binary distribution only includes: 

```
kroxylicious-api-0.4.1.jar
kroxylicious-app-0.4.1.jar
kroxylicious-runtime-0.4.1.jar
```

so there's a bit of work to do to get the KMS/filter dependencies installed.

Reply:
should we make a batteries included distro with all kroxy filters something we offer? otherwise we could create another encryption filter tarball containing.

```
caffeine-3.1.8.jar
kroxylicious-encryption-0.5.0-SNAPSHOT.jar
kroxylicious-kms-0.5.0-SNAPSHOT.jar
```

then the process would be take your clean Kroxy distro, unpack the encryption tarball from mavenCentral/github and download the KMS implementations from mavenCentral/github. A bit of a mission to explain.

Reply:
or instruct them to curl those deps down out of central into their `libs` dir

Reply:
At this stage of  Kroxylicious's life, we have only one filter that has received serious work.  I think most users will be experimenting with it, rather than building their own.  I say with bundle it in the kroxylicious tgz for the time being.  We are already including the filters in the images.  Let's keep things simple.

Once we get to the point where we several filters, we can think alternative packaging strategies.  

WDYAT?   @SamBarker @tombentley

Reply:
>  I say with bundle it in the kroxylicious tgz for the time being.

+1